### PR TITLE
Simple topology caches

### PR DESCRIPTION
**Pull request by deniaa**

### CODE REVIEW

**Comment by deniaa** [comment]
This is a part of this issue: https://github.com/vostok/clusterclient.topology.sd/issues/1 
I believe we can implement some more smarter caching strategies. But this little helpers can be usefull already.

File: Vostok.ClusterClient.Topology.SD/Transforms/KeepLastNonEmptyTopologyTransform.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        {
+            this.log = log ?? LogProvider.Get();
+        }
+
+        public IEnumerable<Uri> Transform(IServiceTopology topology)
+        {
+            var replicas = topology.Replicas;
+            if (replicas.Count == 0 && lastSeenReplicas?.Count > 0)
+            {
+                log.Warn("New observed topology is empty, but we have a cached one. Last seen cached topology will be used.");
```
Comment (iloktionov) [inline Vostok.ClusterClient.Topology.SD/Transforms/KeepLastNonEmptyTopologyTransform.cs:25]: These log messages are going to be spammy (one for each request while there are no replicas in current topology).

File: Vostok.ClusterClient.Topology.SD/Transforms/NeverForgetReplicasTransform.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+using JetBrains.Annotations;
+using Vostok.Commons.Helpers.Topology;
+using Vostok.ServiceDiscovery.Abstractions;
+
+namespace Vostok.Clusterclient.Topology.SD.Transforms
+{
+    [PublicAPI]
+    public class NeverForgetReplicasTransform : IServiceTopologyTransform
+    {
+        private readonly HashSet<Uri> detectedReplicas = new HashSet<Uri>(ReplicaComparer.Instance);
```
Comment (iloktionov) [inline Vostok.ClusterClient.Topology.SD/Transforms/NeverForgetReplicasTransform.cs]: Is it even safe to use `ReplicaComparer` here? I mean, it totally disregards the path portion of the `Uri`, which might be acceptable for Snitch where this comparer once originated, but not correct in general.

File: Vostok.ClusterClient.Topology.SD/ServiceDiscoveryClusterProviderSettings.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+{
+    [PublicAPI]
+    public class ServiceDiscoveryClusterProviderSettings
+    {
+        /// <summary>
+        /// If not set, replicas will be extracted from the IServiceTopology as is.
+        /// Blacklist will be applied in <see cref="ServiceDiscoveryClusterProvider" /> after calling this Transform.
+        /// </summary>
+        [CanBeNull]
+        public IServiceTopologyTransform ServiceTopologyTransform { get; set; }
```
Comment (iloktionov) [inline Vostok.ClusterClient.Topology.SD/ServiceDiscoveryClusterProviderSettings.cs:14]: All of these transforms only retain their internal state within a single `ClusterClient` instance (won't work for clients that use `new Client()` for every request). Was this an explicit design decision?

**Pull request merged**